nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/pubsub/PublishGCPubSub.java

```diff
@@ -0,0 +1,240 @@
+        return ImmutableList.of(PROJECT_ID,
+                GCP_CREDENTIALS_PROVIDER_SERVICE,
+                TOPIC_NAME,
+                EXECUTOR_COUNT,
+                BATCH_SIZE);
+    }
+
+    @Override
+    protected PropertyDescriptor getSupportedDynamicPropertyDescriptor(final String propertyDescriptorName) {
+        return new PropertyDescriptor.Builder()
+                .required(false)
+                .name(propertyDescriptorName)
+                .displayName(propertyDescriptorName)
+                .addValidator(StandardValidators.ATTRIBUTE_KEY_PROPERTY_NAME_VALIDATOR)
+                .dynamic(true)
+                .expressionLanguageSupported(ExpressionLanguageScope.FLOWFILE_ATTRIBUTES)
+                .build();
+    }
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return Collections.unmodifiableSet(
+                new HashSet<>(Arrays.asList(REL_SUCCESS, REL_FAILURE, REL_RETRY))
+        );
+    }
+
+
+    @Override
+    public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+        FlowFile flowFile = session.get();
+
+        if (flowFile == null) {
+            context.yield();
+            return;
+        }
+
+        final long startNanos = System.nanoTime();
+
+        final Map<String, String> attributes = new HashMap<>();
+
+        try {
+            publisher = getPublisherBuilder(context, flowFile).build();
+
+            final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+            session.exportTo(flowFile, baos);
+            final ByteString flowFileContent = ByteString.copyFromUtf8(baos.toString());
+
+            PubsubMessage message = PubsubMessage.newBuilder().setData(flowFileContent)
+                    .setPublishTime(Timestamp.newBuilder().build())
+                    .putAllAttributes(getDynamicAttributesMap(context, flowFile))
+                    .build();
+
+            ApiFuture<String> messageIdFuture = publisher.publish(message);
+
+            while (messageIdFuture.isDone()) {
+                Thread.sleep(1000L);
+            }
+
+            final String messageId = messageIdFuture.get();
+
+            attributes.put(MESSAGE_ID_ATTRIBUTE, messageId);
+            attributes.put(TOPIC_NAME_ATTRIBUTE, getTopicName(context, flowFile).toString());
+
+            flowFile = session.putAllAttributes(flowFile, attributes);
+        } catch (IOException e) {
+            getLogger().error("Routing to 'failure'. Failed to build the Google Cloud PubSub Publisher due to ", e);
+            session.transfer(flowFile, REL_FAILURE);
+            return;
+        } catch (InterruptedException | ExecutionException e) {
+            getLogger().error("Routing to 'retry'. Failed to publish the message to Google PubSub topic due to ", e);
+            session.transfer(flowFile, REL_RETRY);
+            return;
+        } finally {
+            shutdownPublisher();

```

Comment:
Actually I previously had ExpressionLanguageScope for `TOPIC`  set to `VariableRegistry` and in `unScheduled` shutdown the publisher since that would be more efficient but the GCP SDK kept on writing WARN and ERROR log messages saying "shutdown was not called and for proper resource utilization, call shutdown()" That's why I had to go with this. I'll try to check with Google folks and there is this `TransportChannel` API.. have to check if that offers anything of interest in this regard.